src/upstream_temperature_feedback.cxx

```diff
@@ -0,0 +1,60 @@
+#include "../include/upstream_temperature_feedback.hxx"
+
+#include <bout/mesh.hxx>
+using bout::globals::mesh;
+
+void UpstreamTemperatureFeedback::transform(Options& state) {
+  Options& species = state["species"][name];
+
+  // Doesn't need all boundaries to be set
+  Field3D T = getNoBoundary<Field3D>(species["temperature"]);
+
+  auto time = get<BoutReal>(state["time"]);
+
+  for (RangeIterator r = mesh->iterateBndryLowerY(); !r.isDone(); r++) {
+    int jz = 0;
+    error = temperature_upstream - T(r.ind, mesh->ystart, jz);
+
+    // PI controller, using crude integral of the error
+    if (temperature_error_lasttime < 0.0) {
+      // First time this has run
+      temperature_error_lasttime = time;
+      temperature_error_last = error;
+    }
+
+    // Integrate using Trapezium rule
+    if (time > temperature_error_lasttime) { // Since time can decrease
+      temperature_error_integral += (time - temperature_error_lasttime) * 0.5 *
+        (error + temperature_error_last);
+    }
+
+    if ((temperature_error_integral < 0.0) && temperature_integral_positive) {
+      // Limit temperature_error_integral to be >= 0
+      temperature_error_integral = 0.0;
+    }
+
+    // Calculate source from combination of error and integral
+    integral_term = temperature_controller_i * temperature_error_integral;
+    proportional_term = temperature_controller_p * error;
+    source_multiplier = proportional_term + integral_term;
+
+    if ((source_multiplier < 0.0) && temperature_source_positive) {
+      source_multiplier = 0.0; // Don't remove particles
+    }
+
+    temperature_error_last = error;
+    temperature_error_lasttime = time;
+    break;
+  }
```

Comment:
Good question. This might be left in because in a 1D domain there's only one point at the boundary. The loop is iterating in X (radial coordinate) over all point at the lower Y boundary. In 1D simulations only the processor at the lower Y (upstream) end of the domain will enter the loop. It should exit after one iteration, because there's only one point on the boundary, but this `break` is probably there to make sure.

In 2D or 3D simulations we'd have to consider which temperature or density should be taken. Probably the maximum? But in that case the Lower Y boundary would be one of the targets. Probably need some more general strategy if we want to control input power in 2D simulations.